src/main/rx/fport.c

```diff
@@ -0,0 +1,282 @@
+/*
+ * This file is part of Cleanflight.
+ *
+ * Cleanflight is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * Cleanflight is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with Cleanflight.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include <stdbool.h>
+#include <stdint.h>
+#include <stdlib.h>
+
+#include "platform.h"
+
+#ifdef SERIAL_RX
+
+#include "build/debug.h"
+
+#include "common/utils.h"
+
+#include "drivers/time.h"
+
+#include "io/serial.h"
+
+#ifdef TELEMETRY
+#include "telemetry/telemetry.h"
+#endif
+
+#include "rx/rx.h"
+#include "rx/sbus_channels.h"
+#include "rx/fport.h"
+
+
+#define FPORT_TIME_NEEDED_PER_FRAME 5000
+
+#define FPORT_FRAME_PAYLOAD_SIZE_CONTROL (SBUS_CHANNEL_DATA_LENGTH + 3)
+#define FPORT_FRAME_PAYLOAD_SIZE_DOWNLINK 8
+#define FPORT_FRAME_MAX_PAYLOAD_SIZE FPORT_FRAME_PAYLOAD_SIZE_CONTROL 
+
+#define FPORT_FRAME_MARKER 0x7E
+
+#define FPORT_ESCAPE_CHAR 0x7D
+#define FPORT_ESCAPE_MASK 0x20
+
+#define FPORT_CRC_VALUE 0xFF
+
+#define FPORT_BAUDRATE 115200
+
+#define FPORT_PORT_OPTIONS (SERIAL_STOPBITS_1 | SERIAL_PARITY_NO)
+
+enum {
+	DEBUG_FPORT_FRAME_TYPE = 0,
+	DEBUG_FPORT_FRAME_INTERVAL,
+	DEBUG_FPORT_FRAME_ERRORS,
+	DEBUG_FPORT_FRAME_LAST_ERROR,
+};
+
+enum {
+	DEBUG_FPORT_NO_ERROR = 0,
+	DEBUG_FPORT_ERROR_OVERSIZE,
+	DEBUG_FPORT_ERROR_CHECKSUM,
+	DEBUG_FPORT_ERROR_FRAMING,
+	DEBUG_FPORT_ERROR_TIMEOUT,
+	DEBUG_FPORT_ERROR_PAYLOAD_SIZE = 0x10
+};
+
+typedef enum fportFrameState_e {
+	FPORT_FRAME_STATE_IDLE,
+	FPORT_FRAME_STATE_LENGTH,
+	FPORT_FRAME_STATE_PAYLOAD,
+	FPORT_FRAME_STATE_CHECKSUM,
+	FPORT_FRAME_STATE_END_MARKER,
+	FPORT_FRAME_STATE_COMPLETE,
+} fportFrameState_t;
+
+typedef enum fportFrameType_e {
+	FPORT_FRAME_TYPE_CONTROL = 0x00,
+	FPORT_FRAME_TYPE_DOWNLINK = 0x01,
```

Comment:
is downlink correct term? Channel returning back to transmitter is usually called uplink ...

Reply:
This is what FrSky calls it in their specification. (In general, naming in here uses specification where possible).